third_party/move/move-vm/integration-tests/src/tests/function_value_tests.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    compiler::{as_module, compile_units_move_2_2},
+    tests::{
+        execute_function_for_test_and_apply_change_set,
+        execute_function_with_single_storage_for_test,
+    },
+};
+use move_binary_format::{
+    deserializer::DeserializerConfig,
+    file_format_common::{IDENTIFIER_SIZE_MAX, VERSION_8},
+};
+use move_core_types::{account_address::AccountAddress, ident_str, value::MoveValue};
+use move_vm_runtime::{config::VMConfig, RuntimeEnvironment};
+use move_vm_test_utils::InMemoryStorage;
+
+fn storage_with_binary_format_version_8() -> InMemoryStorage {
+    let vm_config = VMConfig {
+        deserializer_config: DeserializerConfig::new(VERSION_8, IDENTIFIER_SIZE_MAX),
+        paranoid_type_checks: true,
+        ..Default::default()
+    };
+    let runtime_environment = RuntimeEnvironment::new_with_config(vec![], vm_config);
+    InMemoryStorage::new_with_runtime_environment(runtime_environment)
+}
+
+#[test]
+fn test_function_value_resolution() {
```

Comment:
Moving this as the following transaction test under `move-compiler-v2/transactional-tests/tests/no-v1-comparison/closures/`:

```
//# publish
module 0x42::test {
    struct Func has key {
            f: |()|u64 has store + copy,
    }

    public fun identity<T>(x: T): T { x }

    public fun init(account: &signer) {
        let f = || 0x42::test::identity(23);
        move_to(account, Func { f });
    }

    public fun call(addr: address) acquires Func {
        let f = &borrow_global<Func>(addr).f;
        let value = (*f)();
        assert!(value == 23, 777);
    }
}

//# run 0x42::test::init --signers 0x42

//# run 0x42::test::call --args @0x42
```

Reply:
Done